zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+            };
+
+            let vote_lock_counter = <Locks<T>>::try_mutate(
+                &sender,
+                |LockInfo(ref mut lock_info)| -> Result<u32, DispatchError> {
+                    match lock_info.binary_search_by_key(&market_id, |i| i.0) {
+                        Ok(i) => {
+                            let prev_highest_amount: BalanceOf<T> = lock_info[i].1;
+                            // concious decision here to not throw an error when the new amount is not higher
+                            // because the user would have to pay worst case fees then
+                            if amount > prev_highest_amount {
+                                let diff = amount.saturating_sub(prev_highest_amount);
+                                add_to_outcome_sum(diff);
+                                lock_info[i].1 = amount;
+                            }
+                        }
+                        Err(i) => {
+                            ensure!(
+                                lock_info.len() as u32 <= T::MaxGlobalDisputeVotes::get(),
+                                Error::<T>::MaxVotesReached
+                            );
+                            add_to_outcome_sum(amount);
+                            lock_info.insert(i, (market_id, amount));
+                        }
+                    }
+                    Ok(lock_info.len() as u32)
+                },
+            )?;
+
+            T::Currency::extend_lock(
+                T::VoteLockIdentifier::get(),
+                &sender,
+                amount,
+                WithdrawReasons::TRANSFER,
+            );
+
+            Self::deposit_event(Event::VotedOnOutcome(market_id, outcome, amount));
+            Ok(Some(T::WeightInfo::vote_on_outcome(outcome_owners_len, vote_lock_counter)).into())
+        }
+
+        #[frame_support::transactional]
+        #[pallet::weight(T::WeightInfo::unlock_vote_balance(
+            T::MaxGlobalDisputeVotes::get(),
+            T::MaxOwners::get()
+        ))]
+        pub fn unlock_vote_balance(
+            origin: OriginFor<T>,
+            voter: AccountIdLookupOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_signed(origin)?;
+            let voter = T::Lookup::lookup(voter)?;
+
+            let mut lock_needed: BalanceOf<T> = Zero::zero();
+            let vote_lock_counter =
+                <Locks<T>>::mutate(&voter, |LockInfo(ref mut lock_info)| -> u32 {
+                    let vote_lock_counter = lock_info.len() as u32;
+                    lock_info.retain(|&(market_id, locked_balance)| {
+                        // weight component MaxOwners comes from querying the winner information
+                        match <Winners<T>>::get(market_id) {
+                            Some(winner_info) if winner_info.is_finished => false,
+                            _ => {
+                                lock_needed = lock_needed.max(locked_balance);
```

Comment:
We have to sum it up here. Otherwise we only unlock the highest lock, but remove all locks for finished markets for that account anyways. So if `finished_locks = [10,20,30]`, we remove all of them but only unlock `30` instead of `60`.

Reply:
Or is the design idea that we only have to lock an amount once and then can vote with it on as many markets we like to?

Reply:
Good thoughts of you.
The rule is: Remove all locks from resolved global disputes. Then find the highest lock from all unresolved global disputes. 

I hope this answers your question.

Reply:
ebbcd381794eed82c7083c8a2c71cf2e46fe1bc6